File: Full Stack Develop/Programming Language/Python/demos/d03_functions.py
"""03 Functions — default parameters, *args/**kwargs, keyword arguments, closures, type annotations.

Key points:
  1. Both parameters and return values can be annotated; Callable describes a "function type".
  2. *args collects positional arguments into a tuple, **kwargs collects keyword arguments into a dict.
  3. Before `/` are positional-only parameters, after `*` are keyword-only parameters.
  4. Default values are evaluated once "at definition time" — a mutable default is a classic trap (see 11).
  5. Functions are first-class objects: assignable, passable, returnable → higher-order functions and closures.
"""

from collections.abc import Callable

from .log import note, show, title


def add(a: int, b: int) -> int:
    return a + b


def greet(name: str, greeting: str = "Hi", *tags: str) -> str:
    """Default parameter + variadic positional arguments."""
    suffix = f" [{', '.join(tags)}]" if tags else ""
    return f"{greeting}, {name}{suffix}"


def make_user(*, name: str, age: int = 0) -> dict[str, object]:
    """Everything after `*` is "keyword-only"; callers must name the arguments."""
    return {"name": name, "age": age}


def multiplier(factor: int) -> Callable[[int], int]:
    """Higher-order function: returns a closure that remembers factor."""
    return lambda n: n * factor


def run() -> None:
    title("03 Functions")

    show("add(2, 3)", add(2, 3))

    note("default parameters / variadic positional arguments *args")
    show("greet('Harry')", greet("Harry"))
    show("greet('Harry', 'Hello')", greet("Harry", "Hello"))
    show("greet('Harry', 'Hey', 'vip', 'new')", greet("Harry", "Hey", "vip", "new"))

    note("keyword-only parameters: callers must name them, which reads better")
    show("make_user(name='Alice', age=30)", make_user(name="Alice", age=30))

    note("**kwargs collects arbitrary keyword arguments into a dict")

    def describe(**props: object) -> str:
        return ", ".join(f"{k}={v}" for k, v in props.items())

    show("describe(a=1, b=2)", describe(a=1, b=2))

    note("closure + lambda: triple remembers factor=3")
    triple = multiplier(3)
    show("triple(10)", triple(10))

    note("functions as arguments: built-in map/filter/sorted all take a function")
    show("sorted(words, key=len)", sorted(["banana", "fig", "apple"], key=len))
    show("list(map(str.upper, ...))", list(map(str.upper, ["a", "b"])))

File: Full Stack Develop/Programming Language/Rust/src/demos/d03_functions.rs
use crate::log::{note, show, title};

/// Functions — expression returns, closures, higher-order functions, generics.
/// Key points:
///   1. The last "expression" in a function body (no semicolon) is the return value.
///   2. Rust has no default parameters/overloading; use Option or a builder to express optionality.
///   3. Closures |x| ...; move transfers ownership of captured environment variables into the closure.
///   4. Use impl Fn / generics to accept functions, implementing higher-order functions.
///   5. Use a tuple for multiple return values.

fn add(a: i32, b: i32) -> i32 {
    a + b // expression return: no semicolon, no return
}

fn divmod(a: i32, b: i32) -> (i32, i32) {
    (a / b, a % b)
}

fn sum(nums: &[i32]) -> i32 {
    nums.iter().sum()
}

// return a closure: impl Fn, move-capturing factor
fn multiplier(factor: i32) -> impl Fn(i32) -> i32 {
    move |n| n * factor
}

// higher-order function: pass behavior in as a parameter
fn apply<T, R>(value: T, f: impl Fn(T) -> R) -> R {
    f(value)
}

pub fn run() {
    title("03 Functions");

    show("add(2, 3)", add(2, 3));

    note("use a tuple for multiple return values");
    show("divmod(17, 5)", divmod(17, 5));

    note("slice parameter &[i32]: zero-copy, compatible with arrays and Vec");
    show("sum(&[1,2,3,4])", sum(&[1, 2, 3, 4]));

    note("closure + move capture");
    let triple = multiplier(3);
    show("triple(10)", triple(10));

    note("higher-order function: a closure as an argument");
    show("apply(9, |n| n*n)", apply(9, |n: i32| n * n));

    note("no default parameters: use Option to express optionality");
    let greet = |name: &str, greeting: Option<&str>| format!("{}, {}", greeting.unwrap_or("Hi"), name);
    show("greet(None)", greet("Harry", None));
    show("greet(Some(\"Hello\"))", greet("Harry", Some("Hello")));
}
